server/src/main/java/org/elasticsearch/plugins/PluginsService.java

```diff
@@ -57,6 +57,20 @@
 
 public class PluginsService implements ReportingService<PluginsAndModules> {
 
+    /**
+     * A loaded plugin is one for which Elasticsearch has successfully constructed an instance of the plugin's class
+     * @param info Metadata about the plugin, usually loaded from plugin properties
+     * @param instance The constructed instance of the plugin's main class
+     */
+    record LoadedPlugin(PluginInfo info, Plugin instance) {}
+
+    /**
+     * A constructed plugin has an associated classloader
+     * @param plugin A constructed instance of a plugin's main class
+     * @param loader The classloader for the plugin
+     */
+    private record PluginClassloaderContext(Plugin plugin, ClassLoader loader) {}
```

Comment:
I think this can be combined into LoadedPlugin, adding the ClassLoader there.

Reply:
I think I managed to correctly add the ClassLoader to the LoadedPlugin object. The main thing I am not sure about is that I needed to pick something arbitrary for the classpath plugins, so I just used the current class's classloader.

Reply:
Changed my mind on this -- I used `null` for classpath plugin classloaders and added an `assert` statement to verify we don't hit the bad case when building classloaders for extended plugins.